OA-PASS/pass-docker
# Notification Services Integration

## About
This PR adds Notification Services (NS) to `pass-docker`.  NS will respond to the creation of `SubmissionEvent` resources in Fedora by dispatching notification emails to the intended recipients.

Because Ember is not creating `SubmissionEvent` resources at this time, the only way to see this PR in action is to execute an [IT from the notification-services repo](https://github.com/OA-PASS/notification-services/blob/master/notification-integration/src/test/java/org/dataconservancy/pass/notification/NotificationSmokeIT.java) against services running in `pass-docker`.

Since there has not been an official release of `notification-services`, currently the executable jar is included in this PR.

## Test Instructions

1. Start Fedora, mail, and ldap in the background:
    - `docker-compose up -d ldap mail fcrepo`

2. Wait for Mail to start:
    - `docker logs -f mail`

```
   ...
    Oct 30 13:45:06 mail postfix/master[1428]: daemon started -- version 3.1.8, configuration /etc/postfix
```

3. Wait for Fedora to start:
    - `docker logs -f fcrepo`

```
    ...
    INFO [main] org.apache.catalina.startup.Catalina.start Server startup in 60508 ms
```

4. Start notification services in foreground:
    - `docker-compose up notification`

```
    15:01:36.821 [                main] INFO  [               NotificationApp] - >>>> Starting Notification Services (version: 0.0.1-3.1-SNAPSHOT branch: master commit: ba3fdc3d commit date: 2018-10-30T10:57:24-0400)
    ....
    15:01:41.613 [                main] INFO  [             SpringApplication] - No active profile set, falling back to default profiles: default
```

5. In a separate terminal window, check out `ba3fdc3` from the notification-services `master` branch.

6. Install the project without executing tests:
    - `mvn -U install -Dspring-boot.stop.skip -Dspring-boot.run.skip -Ddocker.skip=true -DskipTests=true -Dmaven.javadoc.skip=true -B -V`

7. Enter the `notification-integration` module, and run the `NotificationSmokeIT`, but use the services running in docker:

```
mvn clean verify -Dit.test=NotificationSmokeIT \
                 -Dspring-boot.stop.skip -Dspring-boot.run.skip \
                 -Dldap.skip -Dfcrepo.skip -Dmail.skip \
                 -Dmail.server=<docker host> -Dfcrepo.server=<docker host> -Dmail.server=<docker host> \
                 -Dfcrepo.http.port=8080 \
                 -Djms.host=activemq -Djms.port=61616
```

8. Wait for the IT to finish
    - note: there is a 20 second "sleep" in the IT at the end that gives time for things to complete
    - this will be improved in the future by adding an aspect that exposes some of the internals of NS, allowing the IT to include code-based verifications instead of userland checks.

9. Check for new mail, and examine content with respect to the templates
    - `docker exec mail ls /var/mail/jhu.edu/staffWithGrants/new`
    - `docker exec mail cat /var/mail/jhu.edu/staffWithGrants/new/<filename>`
    - `docker exec mail ls /var/mail/jhu.edu/notification-demo-cc/new`
    - `docker exec mail cat /var/mail/jhu.edu/notification-demo-cc/new/<filename>`
